src/cli/cli.cpp

```diff
@@ -8251,256 +8251,357 @@ otError Interpreter::ProcessCommand(Arg aArgs[])
     }
 
     static constexpr Command kCommands[] = {
-#if OPENTHREAD_FTD || OPENTHREAD_MTD
-#if OPENTHREAD_CONFIG_BORDER_AGENT_ENABLE
+#ifdef CMD_ba
         CmdEntry("ba"),
 #endif
-#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+#ifdef CMD_bbr
         CmdEntry("bbr"),
 #endif
-#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+#ifdef CMD_br
         CmdEntry("br"),
 #endif
+#ifdef CMD_bufferinfo
         CmdEntry("bufferinfo"),
+#endif
+#ifdef CMD_ccathreshold
         CmdEntry("ccathreshold"),
-#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+#endif
+#ifdef CMD_ccm
         CmdEntry("ccm"),
 #endif
+#ifdef CMD_channel
         CmdEntry("channel"),
-#if OPENTHREAD_FTD
+#endif
+#ifdef CMD_child
         CmdEntry("child"),
+#endif
+#ifdef CMD_childip
         CmdEntry("childip"),
+#endif
+#ifdef CMD_childmax
         CmdEntry("childmax"),
+#endif
+#ifdef CMD_childrouterlinks
         CmdEntry("childrouterlinks"),
 #endif
+#ifdef CMD_childsupervision
         CmdEntry("childsupervision"),
+#endif
+#ifdef CMD_childtimeout
         CmdEntry("childtimeout"),
-#if OPENTHREAD_CONFIG_COAP_API_ENABLE
+#endif
+#ifdef CMD_coap
         CmdEntry("coap"),
 #endif
-#if OPENTHREAD_CONFIG_COAP_SECURE_API_ENABLE
+#ifdef CMD_coaps
         CmdEntry("coaps"),
 #endif
-#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE
+#ifdef CMD_coex
         CmdEntry("coex"),
 #endif
-#if OPENTHREAD_CONFIG_COMMISSIONER_ENABLE && OPENTHREAD_FTD
+#ifdef CMD_commissioner
         CmdEntry("commissioner"),
 #endif
-#if OPENTHREAD_FTD
+#ifdef CMD_contextreusedelay
         CmdEntry("contextreusedelay"),
 #endif
+#ifdef CMD_counters
         CmdEntry("counters"),
-#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+#endif
+#ifdef CMD_csl
         CmdEntry("csl"),
 #endif
+#ifdef CMD_dataset
         CmdEntry("dataset"),
+#endif
+#ifdef CMD_debug
         CmdEntry("debug"),
-#if OPENTHREAD_FTD
+#endif
+#ifdef CMD_delaytimermin
         CmdEntry("delaytimermin"),
 #endif
+#ifdef CMD_detach
         CmdEntry("detach"),
-#endif // OPENTHREAD_FTD || OPENTHREAD_MTD
-#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_MLE_DEVICE_PROPERTY_LEADER_WEIGHT_ENABLE
+#endif
+#ifdef CMD_deviceprops
         CmdEntry("deviceprops"),
 #endif
-#if OPENTHREAD_CONFIG_DIAG_ENABLE
+#ifdef CMD_diag
         CmdEntry("diag"),
 #endif
-#if OPENTHREAD_FTD || OPENTHREAD_MTD
+#ifdef CMD_discover
         CmdEntry("discover"),
-#if OPENTHREAD_CONFIG_DNS_CLIENT_ENABLE || OPENTHREAD_CONFIG_DNSSD_SERVER_ENABLE || \
-    OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+#endif
+#ifdef CMD_dns
         CmdEntry("dns"),
 #endif
-#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+#ifdef CMD_domainname
         CmdEntry("domainname"),
 #endif
-#if OPENTHREAD_CONFIG_DUA_ENABLE
+#ifdef CMD_dua
```

Comment:
I like the idea of adding `cmd_condition_eval.h` so we can organize commands into groups depending on the related `CONFIG` under which they are used.

However, the fact that we need to add `#ifdef CMD_xyz` to every item here makes it complex.

It would be great if we could somehow hide this within the `CmdEntry()` itself. Something like if the `CMD_xyz` is not enabled, then `CmdEntry()` becomes empty/null... I don't have any immediate suggestions on how we may be able to do this, though? (We may be able to play with macros within macros, but the ugly part is how to deal with commas).

Reply:
This is an interesting problem and I spent the weekend researching on it 😆 

First, unfortunately, we cannot use `#ifdef` in since a marco, like:
```
#define ENTRY(name) \
    #ifdef name \
        ...     \
    #endif
```

Then I tried to use template to solve this problem. The idea is to use template specialization and make a comamnd table type which accepts all the commands in the template argument. And finally do a reinterpret cast. But I still got one unresolvable issue: In `static_assert(BinarySearch::IsSorted(kCommands))`, I didn't find any way to successfully reinterpret_cast the type.

My prototype code:
```
template <typename... Args> struct CmdList;

template <typename T, typename... Args> struct CmdList<T, Args...>
{
    using Other                  = CmdList<Args...>;
    static constexpr uint8_t kLen = T::kLen + Other::kLen;

    T     conditionalCmd;
    Other other;
};

template <> struct CmdList<>
{
    static constexpr uint8_t kLen = 0;
};

#define CMD_ENTRY_SPEC(aCommandName) \
template <> struct Interpreter::ConditionalCmdEntry<Cmd(aCommandName)> \
{ \
    static constexpr uint8_t kLen = 1; \
    Command mCmd = {aCommandName, &Interpreter::Process<Cmd(aCommandName)>}; \
};

CMD_ENTRY_SPEC("ba")
CMD_ENTRY_SPEC("version")
```

The default template for `ConditionalCmdEntry` is an empty one which doesn't contain any member. And we define those commands not required as "". In this way they will become the default template. So finally `CmdList` will be bitwisely same as a command entry list. But we need a reinterpret_cast.

In `ProcessCommand`:
```
    using CommandsTable = CmdList<
        ConditionalCmdEntry<Cmd(CMD_ba)>,
        ConditionalCmdEntry<Cmd(CMD_version)>>;

    static constexpr CommandsTable kCommands_;

    static_assert(BinarySearch::IsSorted(...), "Command Table is not sorted");
```
However I don't find any way to do the cast successfully in `static_assert`. So I don't get any better options for enhancement here.

---

As another topic, I found an issue in cli: the `CommandId` is uint64_t. In the encoding function `Cmd`, one character will take one byte. That means a CommandId can only encode to a string with at most 8 chars. Here we have some cli commands that are much longer, for example, `contextreusedelay`. I understand that even we simply apply `Cmd` to do the encoding, it is very unlikely that we will get any conflicts in the encoding. But is it by design?

Because it just come up with me that, if we don't store `const char *` in each command entry and store an unique command Id instead, we can save much memory. When we compare, it will be even much more efficient. The only problem is that, `AreStringInOrder` won't work if the strings are longer than 8 bytes. Otherwise, we can still implement `AreStringInOrder` by a similar recursive approach.